components/script/dom/htmlelement.rs

```diff
@@ -839,10 +904,97 @@ impl VirtualMethods for HTMLElement {
                     DOMString::from(&**attr.value()),
                 );
             },
+            (&local_name!("form"), mutation) => {
+                if self.is_form_associated_custom_element() {
+                    self.form_attribute_mutated(mutation);
+                }
+            },
+            (&local_name!("disabled"), AttributeMutation::Set(_)) => {
+                // Adding a "disabled" attribute disables an enabled form element.
+                if self.is_form_associated_custom_element() {
+                    let el = self.upcast::<Element>();
+                    if el.enabled_state() {
+                        el.set_disabled_state(true);
+                        el.set_enabled_state(false);
+                        ScriptThread::enqueue_callback_reaction(
+                            el,
+                            CallbackReaction::FormDisabled(true),
+                            None,
+                        );
+                    }
+                }
+            },
+            (&local_name!("disabled"), AttributeMutation::Removed) => {
+                // Removing the "disabled" attribute may enable a disabled
+                // form element, but a fieldset ancestor may keep it disabled.
+                if self.is_form_associated_custom_element() {
+                    let el = self.upcast::<Element>();
+                    if el.disabled_state() {
+                        el.set_disabled_state(false);
+                        el.set_enabled_state(true);
+                        el.check_ancestors_disabled_state_for_form_control();
+                        if el.enabled_state() {
+                            ScriptThread::enqueue_callback_reaction(
+                                el,
+                                CallbackReaction::FormDisabled(false),
+                                None,
+                            );
+                        }
+                    }
+                }
+            },
             _ => {},
         }
     }
 
+    fn bind_to_tree(&self, context: &BindContext) {
+        if let Some(ref s) = self.super_type() {
+            s.bind_to_tree(context);
+        }
+        let el = self.upcast::<Element>();
+        el.update_sequentially_focusable_status();
+
+        // Binding to a tree can disable a form control if one of the new
+        // ancestors is a fieldset.
+        if self.is_form_associated_custom_element() {
+            if el.enabled_state() {
```

Comment:
No need to nest these conditions:

```suggestion
        if self.is_form_associated_custom_element() && el.enabled_state() {
```